File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
import sys
input = sys.stdin.readline

def DFS(l, tot, plus, minus, mul, div):
    global small, big
    if l == n:
        small = min(small, tot)
        big = max(big, tot)
        return

    if plus:
        DFS(l+1, tot+li[l], plus-1, minus, mul, div)
    if minus:
        DFS(l+1, tot-li[l], plus, minus-1, mul, div)
    if mul:
        DFS(l+1, tot*li[l], plus, minus, mul-1, div)
    if div:
        DFS(l+1, int(tot/li[l]), plus, minus, mul, div-1)

n = int(input())
li = list(map(int, input().split()))
a,b,c,d = map(int, input().split())
small = 1e9
big = -1e9

DFS(1,li[0],a,b,c,d)
print(big)
print(small)

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	static int N, pl, mi, mul, div, big, small;
	static int[] numbers;
	public static void main(String[] args) throws NumberFormatException, IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		
		numbers = new int[N];
		StringTokenizer st = new StringTokenizer(br.readLine());
		for(int i = 0; i<N; i++) {
			numbers[i] = Integer.parseInt(st.nextToken());
		}
		
		st = new StringTokenizer(br.readLine());
		pl = Integer.parseInt(st.nextToken());
		mi = Integer.parseInt(st.nextToken());
		mul = Integer.parseInt(st.nextToken());
		div = Integer.parseInt(st.nextToken());
		//입력 완료
		
		small = Integer.MAX_VALUE;
		big = Integer.MIN_VALUE;
		dfs(1, numbers[0], pl, mi, mul, div);
		
		System.out.println(big);
		System.out.println(small);
		
	}

	private static void dfs(int cnt, int res, int plus, int minus, int multiple, int divide) {
		
		if(cnt == N) {
			small = Math.min(small, res);
			big = Math.max(big, res);
			return;
		}
		
		if(plus>0) {
			dfs(cnt+1, res+numbers[cnt], plus-1, minus, multiple, divide);
		}
		
		if(minus>0) {
			dfs(cnt+1, res-numbers[cnt], plus, minus-1, multiple, divide);
		}
		
		if(multiple>0) {
			dfs(cnt+1, res*numbers[cnt], plus, minus, multiple-1, divide);
		}
		
		if(divide>0) {
			dfs(cnt+1, res/numbers[cnt], plus, minus, multiple, divide-1);
		}
	}
	
}
